### PR TITLE
Support forward-mode AD

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
Support forward-mode AD and fix a bug when multi-output is needed

### CODE REVIEW

File: deepxde/data/pde_operator.py [hunk 1]
```diff
@@ -237,49 +237,84 @@ def __init__(
...
         bcs_start = np.cumsum([0] + self.pde.num_bcs)
 
-        losses = []
-        for i in range(num_func):
-            out = outputs[i]
-            # Single output
-            if bkd.ndim(out) == 1:
-                out = out[:, None]
-            f = []
+        if config.autodiff == "forward": # forward mode AD
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:243]: Use reverse mode in the `if`, forward mode in the `else`.

File: deepxde/gradients/gradients_forward.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def grad_fn(x):
                 # which is also a tuple. This is useful for further computation, e.g.,
                 # Hessian.
                 self.J[i, j] = (
-                    self.J[j][0][:, i : i + 1],
+                    self.J[j][0][..., i : i + 1],
```
Comment (lululxvi) [inline deepxde/gradients/gradients_forward.py:97]: Why change this?

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_forward.py:97]: fix a bug when multi-output is needed, for example, if output shape is (# of source term, # of coordinates, # of solution), the slice here is wrong, what we really want is slicing in the last dim.

File: deepxde/data/pde_operator.py (hunk 1 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: fix

File: deepxde/gradients/gradients_forward.py (hunk 2 above)
Comment (lululxvi) [reply deepxde/gradients/gradients_forward.py:97]: If I remember correctly, we always reshape the network output to be (batch size, d). Could you point out the code of output shape (batch size, # of coordinates, dim of solution)?

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_forward.py:97]: I find the problem from this case: https://github.com/lululxvi/deepxde/blob/master/examples/operator/stokes_aligned_pideeponet.py

when setting the `num_output=3` and `multi_output_strategy="independent"`, the output shape will be (batch size, # of coordinates, 3).

Comment (lululxvi) [inline deepxde/gradients/gradients_forward.py:97]: Make this change as a separate PR.

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_forward.py]: ok

File: deepxde/data/pde_operator.py [hunk 3]
```diff
@@ -237,49 +237,84 @@ def __init__(
...
+                        aux_var=model.net.auxiliary_vars[i][:, None],
+                    )
+                    losses_i.append(loss_fn(bkd.zeros_like(error), error))
+
+                losses.append(losses_i)
+
+            losses = zip(*losses)
+            # Use stack instead of as_tensor to keep the gradients.
+            losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+        else: # forward mode AD
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:276]:
```
elif config.autodiff == "forward":
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: sure

File: deepxde/data/pde_operator.py [hunk 4]
```diff
@@ -237,49 +237,83 @@ def __init__(
...
         bcs_start = np.cumsum([0] + self.pde.num_bcs)
 
-        losses = []
-        for i in range(num_func):
-            out = outputs[i]
-            # Single output
-            if bkd.ndim(out) == 1:
-                out = out[:, None]
-            f = []
+        if config.autodiff == "reverse": # reverse mode AD
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:243]:
```
losses = []
if config.autodiff == "reverse": # reverse mode AD
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:244]: Ok

File: deepxde/data/pde_operator.py [hunk 5]
```diff
@@ -237,49 +237,82 @@ def __init__(
...
         bcs_start = np.cumsum([0] + self.pde.num_bcs)
-
         losses = []
-        for i in range(num_func):
-            out = outputs[i]
-            # Single output
-            if bkd.ndim(out) == 1:
-                out = out[:, None]
-            f = []
+
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:243]: Delete this empty line.

File: deepxde/data/pde_operator.py [hunk 6]
```diff
@@ -237,49 +237,81 @@ def __init__(
...
-                    aux_var=model.net.auxiliary_vars[i][:, None],
-                )
-                losses_i.append(loss_fn(bkd.zeros_like(error), error))
-
-            losses.append(losses_i)
-
-        losses = zip(*losses)
-        # Use stack instead of as_tensor to keep the gradients.
-        losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+            bcs_start = np.cumsum([0] + self.pde.num_bcs)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:284]: Delete this line, which is already in line 241.

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: Ok

File: deepxde/data/pde_operator.py [hunk 7]
```diff
@@ -237,49 +237,80 @@ def __init__(
...
-
-            losses.append(losses_i)
-
-        losses = zip(*losses)
-        # Use stack instead of as_tensor to keep the gradients.
-        losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]  # noqa
+            # BC
+            for k, bc in enumerate(self.pde.bcs):
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:287]: The BC code is exactly the same for forward and inverse modes. Refactor the code to reduce the redundancy.

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: completed

File: deepxde/data/pde_operator.py [hunk 8]
```diff
@@ -237,49 +237,67 @@ def __init__(
...
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]  # noqa
+        # BC loss
+        for k, bc in enumerate(self.pde.bcs):
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:274]: Use the same variable names as in the original code, if possible.

File: deepxde/data/pde_operator.py [hunk 9]
```diff
@@ -237,49 +237,67 @@ def __init__(
...
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]  # noqa
+        # BC loss
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:273]: Add an empty line before `# BC loss`

File: deepxde/data/pde_operator.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:274]: these variable names are exactly the same as the original code

File: deepxde/data/pde_operator.py (hunk 9 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:273]: ok

File: deepxde/data/pde_operator.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: ok

File: deepxde/data/pde_operator.py [hunk 10]
```diff
@@ -237,49 +237,68 @@ def __init__(
...
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]  # noqa
+
+        # BC loss
+        for j, bc in enumerate(self.pde.bcs):
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:275]: Instead of first loop over self.pde.bcs and then num_func, we can also first loop over num_func, as the original code. It is not clear which one is faster. Could you test?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:275]: I just tested, first loop over bcs is faster

Comment (lululxvi) [reply deepxde/data/pde_operator.py:275]: By how much? Any numbers?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:275]: image
first loop over bcs is 0.005619049072265625s
first loop over func is 0.007124185562133789s

Comment (lululxvi) [reply deepxde/data/pde_operator.py:275]: What is the code of loop over func you tested?

File: deepxde/data/pde_operator.py [hunk 11]
```diff
@@ -237,49 +237,68 @@ def __init__(
...
                 )
-                losses_i.append(loss_fn(bkd.zeros_like(error), error))
-
-            losses.append(losses_i)
-
-        losses = zip(*losses)
-        # Use stack instead of as_tensor to keep the gradients.
-        losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+                error.append(error_i)
+            error = bkd.stack(error, axis=0)  # noqa
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:291]:
```
error = bkd.stack(error, 0)
```

File: deepxde/data/pde_operator.py [hunk 12]
```diff
@@ -237,49 +237,68 @@ def __init__(
...
                     f = [f]
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]  # noqa
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:272]: why do we need `# noqa`?

File: deepxde/data/pde_operator.py (hunk 11 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: ok

File: deepxde/data/pde_operator.py (hunk 12 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: deleted

File: deepxde/data/pde_operator.py [hunk 13]
```diff
@@ -237,49 +237,68 @@ def __init__(
...
+                out = outputs[i]
+                # Single output
+                if bkd.ndim(out) == 1:
+                    out = out[:, None]
+                f = []
+                if self.pde.pde is not None:
+                    f = self.pde.pde(inputs[1], out, model.net.auxiliary_vars[i][:, None])
+                    if not isinstance(f, (list, tuple)):
+                        f = [f]
+                error_f = [fi[bcs_start[-1]:] for fi in f]
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:255]: Reformat the code using black https://github.com/psf/black .

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: completed

File: deepxde/data/pde_operator.py [hunk 14]
```diff
@@ -237,49 +237,74 @@ def __init__(
...
+            # Use stack instead of as_tensor to keep the gradients.
+            losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+        elif config.autodiff == "forward":  # forward mode AD
+
+            def forward_call(trunk_input):
+                return aux[0]((inputs[0], trunk_input))
+
             if self.pde.pde is not None:
-                f = self.pde.pde(inputs[1], out, model.net.auxiliary_vars[i][:, None])
+                f = self.pde.pde(
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:270]: Add a comment here for the dim of f.

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:271]: All done

File: deepxde/data/pde_operator.py [hunk 15]
```diff
@@ -237,49 +237,75 @@ def __init__(
...
+            # Use stack instead of as_tensor to keep the gradients.
+            losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+        elif config.autodiff == "forward":  # forward mode AD
+
+            def forward_call(trunk_input):
+                return aux[0]((inputs[0], trunk_input))
+
             if self.pde.pde is not None:
-                f = self.pde.pde(inputs[1], out, model.net.auxiliary_vars[i][:, None])
+                # f has the shape (`N1`, `N2`), inputs[1] has the shape (`N2`, `dim2`), outputs has the shape (`N1`, `N2`)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:270]:
```
# Each f has the shape (`N1`, `N2`)
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: ok

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: Delete `inputs[1] has the  ...`

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: sure

File: deepxde/data/pde_operator.py [hunk 16]
```diff
@@ -237,49 +237,75 @@ def __init__(
...
+            # Use stack instead of as_tensor to keep the gradients.
+            losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+        elif config.autodiff == "forward":  # forward mode AD
+
+            def forward_call(trunk_input):
+                return aux[0]((inputs[0], trunk_input))
+
             if self.pde.pde is not None:
-                f = self.pde.pde(inputs[1], out, model.net.auxiliary_vars[i][:, None])
+                # Each f has the shape (`N1`, `N2`)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:270]:
```
# Each f has the shape (N1, N2)
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: Ok

File: deepxde/data/pde_operator.py [hunk 17]
```diff
@@ -237,49 +237,75 @@ def __init__(
...
+
+            losses = zip(*losses)
+            # Use stack instead of as_tensor to keep the gradients.
+            losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+        elif config.autodiff == "forward":  # forward mode AD
+
+            def forward_call(trunk_input):
+                return aux[0]((inputs[0], trunk_input))
+
             if self.pde.pde is not None:
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:269]: Add `f = []`

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:270]: added

File: deepxde/data/pde_operator.py [hunk 18]
```diff
@@ -237,49 +237,76 @@ def __init__(
...
                 if not isinstance(f, (list, tuple)):
                     f = [f]
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1] :] for fi in f]
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:277]: Add a comment:
```
Each error has the shape (N1, ~N2)
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:277]: added

File: deepxde/data/pde_operator.py [hunk 19]
```diff
@@ -237,49 +237,77 @@ def __init__(
...
                 )
-                losses_i.append(loss_fn(bkd.zeros_like(error), error))
-
-            losses.append(losses_i)
-
-        losses = zip(*losses)
-        # Use stack instead of as_tensor to keep the gradients.
-        losses = [bkd.reduce_mean(bkd.stack(loss, 0)) for loss in losses]
+                error.append(error_i)
+            error = bkd.stack(error, 0)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:298]: Cannot stack the errors, as different BC errors may have different dim.

```
loss = []
for i in range(num_func):
    ...
    error = bc.error(...)
    loss.append(loss_fn(bkd.zeros_like(error), error))
losses.append(bkd.reduce_mean(bkd.stack(loss, 0)))
```

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:298]: different BC does may have different dim, but my code here first loop over bc and reset error in each loop. so, I think stacking error is doable.

Comment (lululxvi) [reply deepxde/data/pde_operator.py:298]: See updated comment above.

File: deepxde/data/pde_operator.py [hunk 20]
```diff
@@ -237,49 +237,77 @@ def __init__(
...
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
-            for j, bc in enumerate(self.pde.bcs):
-                beg, end = bcs_start[j], bcs_start[j + 1]
-                # The same BC points are used for training and testing.
-                error = bc.error(
+            error_f = [fi[:, bcs_start[-1] :] for fi in f]
+            # Each error has the shape (N1, ~N2)
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:279]: The same thing here. For example, if `loss_fn` is `sum`, then the code is not correct. Need to first compute the loss for each function, then average over functions.

File: deepxde/data/pde_operator.py (hunk 19 above)
Comment (lululxvi) [reply deepxde/data/pde_operator.py]: Is there any performance drop by the new code?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: I just dont understand why `loss_fn` is sum, then `loss_fn(bkd.zeros_like(error), error)` is not correct

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: Assume there are two functions, and for each function, the loss is [1,1,1]. Then the result of original code is 1+1+1+1+1+1 = 6. But the correct way is for each function, the loss is 1+1+1=3, and then the total loss is always average over functions, so it is (3 +3) / 2 = 3.

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: So, the `average` is implicit in any loss function? what If one user indeed want to use a loss function without average?

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: What is the speed of the new old vs old code? I guess the new old might be a little slower.

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: image

old code is faster

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: Then how about the speed of the original code of loop over functions first?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: image

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: So, if we consider both the correctness and speed, the original code of loop over functions first is the best?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: yes

Comment (lululxvi) [reply deepxde/data/pde_operator.py]: Then let us use the original code.

File: deepxde/data/pde_operator.py [hunk 21]
```diff
@@ -265,21 +294,23 @@ def _losses(self, outputs, loss_fn, inputs, model, num_func):
                     end,
                     aux_var=model.net.auxiliary_vars[i][:, None],
                 )
-                losses_i.append(loss_fn(bkd.zeros_like(error), error))
+                error_i.append(loss_fn(bkd.zeros_like(error), error))
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:297]: Use `losses_i` to replace `error_i`

File: deepxde/data/pde_operator.py [hunk 22]
```diff
@@ -237,26 +237,55 @@ def __init__(
         self.train_next_batch()
         self.test()
 
-    def _losses(self, outputs, loss_fn, inputs, model, num_func):
+    def _losses(self, outputs, loss_fn, inputs, model, num_func, aux=None):
         bcs_start = np.cumsum([0] + self.pde.num_bcs)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:241]: keep the empty line 242

File: deepxde/data/pde_operator.py [hunk 23]
```diff
@@ -237,26 +237,55 @@ def __init__(
...
+                )
                 if not isinstance(f, (list, tuple)):
                     f = [f]
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
+            error_f = [fi[:, bcs_start[-1] :] for fi in f]
+            # Each error has the shape (N1, ~N2)
+            losses = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
+        # BC loss
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:280]: Add an empty line before `# BC loss`

File: deepxde/data/pde_operator.py (hunk 21 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:297]: ok

File: deepxde/data/pde_operator.py (hunk 22 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:241]: sure

File: deepxde/data/pde_operator.py (hunk 23 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:280]: sure

File: deepxde/data/pde_operator.py (hunk 20 above)
Comment (lululxvi) [reply deepxde/data/pde_operator.py:280]: Also this part. What is the speed of the current code vs the correct code?

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:280]: image

image

File: deepxde/data/pde_operator.py [hunk 24]
```diff
@@ -237,26 +237,61 @@ def __init__(
...
+                f = self.pde.pde(
+                    inputs[1], (outputs, forward_call), model.net.auxiliary_vars
+                )
                 if not isinstance(f, (list, tuple)):
                     f = [f]
-            error_f = [fi[bcs_start[-1] :] for fi in f]
-            losses_i = [loss_fn(bkd.zeros_like(error), error) for error in error_f]
-
+            error_f = [fi[:, bcs_start[-1]:] for fi in f]
+            # Each error has the shape (N1, ~N2)
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:279]:
```
# Each error has the shape (N1, ~N2)
error_f = [fi[:, bcs_start[-1]:] for fi in f]
```

File: deepxde/data/pde_operator.py [hunk 25]
```diff
@@ -237,26 +237,61 @@ def __init__(
...
+            for error in error_f:
+                error_i = []
+                for i in range(error.shape[0]):
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:282]: `error.shape[0]` --> `bkd.shape(error)[0]`

File: deepxde/data/pde_operator.py [hunk 26]
```diff
@@ -237,26 +237,61 @@ def __init__(
...
+
+        # BC loss
+        error_bc = []
+        for i in range(num_func):
+            losses_i = []
+            out = outputs[i]
+            if bkd.ndim(out) == 1:
+                out = out[:, None]
             for j, bc in enumerate(self.pde.bcs):
                 beg, end = bcs_start[j], bcs_start[j + 1]
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:294]: keep the comment in line 259 in the original code.

File: deepxde/data/pde_operator.py (hunk 25 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: Ok

File: deepxde/data/pde_operator.py (hunk 26 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:294]: added

File: deepxde/data/pde_operator.py (hunk 24 above)
Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py:278]: Ok

File: deepxde/data/pde_operator.py [hunk 27]
```diff
@@ -237,23 +237,59 @@ def __init__(
...
+            # Each error has the shape (N1, ~N2)
+            error_f = [fi[:, bcs_start[-1] :] for fi in f]
+            for error in error_f:
+                error_i = []
+                for i in range(num_func):
+                    error_i.append(loss_fn(bkd.zeros_like(error[i]), error[i]))
+                losses.append(bkd.reduce_mean(bkd.stack(error_i, 0)))
+
+        # BC loss
+        error_bc = []
```
Comment (lululxvi) [inline deepxde/data/pde_operator.py:287]: `error_bc` --> `losses_bc`

Comment (Jerry-Jzy) [reply deepxde/data/pde_operator.py]: ok
